File: cpp/131_PalindromePartitioning.cpp
class Solution {
public:
    vector<vector<string>> palindromes;
    vector<vector<string>> partition(string s) {
        vector<string> curr;
        helper(curr, s, 0);
        return palindromes;
    }
    
    void helper(vector<string>& curr, string s, int start) {
        if (start >= s.length()) {
            palindromes.push_back(curr);
            return;
        }
        
        for (int i = start; i < s.length(); ++i) {
            if (isPalindrome(s, start, i)) {
                curr.push_back(s.substr(start, i - start + 1));
                helper(curr, s, i + 1);
                curr.pop_back();
            }
        }
    }
    
    bool isPalindrome(string s, int start, int end) {
        while (start < end) {
            if (s[start++] != s[end--]) return false;
        }
        return true;
    }
};

File: python/131_PalindromePartitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        def helper(start, curr):
            if start >= len(s):
                partitions.append(curr)
                return
            
            for i in range(start, len(s)):
                if s[start:i+1] == s[start:i+1][::-1]:
                    helper(i + 1, curr + [s[start:i+1]])
        
        partitions = []
        helper(0, [])
        return partitions
